KAFKA-18465: Remove MetadataVersions older than 3.0-IV1

ijuma opened the pull request:
Apache Kafka 4.0 will only support KRaft and 3.0-IV1 is the minimum version supported by KRaft. So, we can assume that Apache Kafka 4.0 will only communicate with brokers that are 3.0-IV1 or newer.

Note that KRaft was only marked as production-ready in 3.3, so we could go further and set the baseline to 3.3. I think we should have that discussion, but it made sense to start with the non controversial parts.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

chia7712 commented on clients/src/test/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequestTest.java:
```diff
@@ -40,20 +40,4 @@ public void testForConsumerRequiresVersion3() {
-            if (version < 3)
-                assertEquals(OffsetsForLeaderEpochRequest.DEBUGGING_REPLICA_ID, parsed.replicaId());
-            else
-                assertEquals(replicaId, parsed.replicaId());
```
Should we keep the test for version=4 ?

ijuma replied:
The value of this test is a bit low, but I added it back.

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java:
```diff
@@ -731,7 +731,6 @@ private void cancelClassicGroupSizeCounter() {
     public void onLoaded(MetadataImage newImage) {
         MetadataDelta emptyDelta = new MetadataDelta(newImage);
         groupMetadataManager.onNewMetadataImage(newImage, emptyDelta);
-        offsetMetadataManager.onNewMetadataImage(newImage, emptyDelta);
```
I guess we don't need to clean up `GroupMetadataManager.scala`, as 4.0 uses a new coordinator by default?

ijuma replied:
I updated that class too where it made sense. Did you spot additional areas of clean-up?

chia7712 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTestContext.java:
```diff
@@ -238,7 +238,7 @@ public static CoordinatorRecord newGroupMetadataRecord(
             ),
             new ApiMessageAndVersion(
                 value,
-                metadataVersion.groupMetadataValueVersion()
```
Could you please remove variable `MetadataVersion metadataVersion` as it is unused now?

ijuma replied:
Done.

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorRecordHelpers.java:
```diff
@@ -552,18 +551,14 @@ public static CoordinatorRecord newEmptyGroupMetadataRecord(
-        MetadataVersion metadataVersion
+        OffsetAndMetadata offsetAndMetadata
     ) {
-        short version = metadataVersion.offsetCommitValueVersion(offsetAndMetadata.expireTimestampMs.isPresent());
```
It seems we need to use version=1 in order to keep "expireTimestamp". otherwise, in v3 `expireTimestamp` will be ignored in serialization.

ijuma replied:
Yeah, this change is wrong - a few things will be reverted as a result.

ijuma replied:
Fixed this one - still working through a few other things.

chia7712 reviewed (commented):
@ijuma nice cleanup!

ijuma commented:
Thanks for the quick review @chia7712. I should have left a note, but I'm still working on this PR. I'll take a look at your comments soon.

ijuma committed:
KAFKA-18465: Remove MetadataVersion methods for versions older than 3.0

Apache Kafka 4.0 will only support KRaft and 3.0 was the first release where KRaft was in a reasonable state (even if not production ready yet). So, we can assume that Apache Kafka 4.0 will only communicate with brokers that are 3.0 or newer.

A couple of notes:

1. KRaft was only marked as production-ready in 3.3, so we could go further and set the baseline to 3.3. I think we should have that discussion, but it made sense to start with the non controversial parts.
2. I have not removed the actual IBP definitions - leaving that for a separate PR.

ijuma committed:
Remove IBP versions older than 3.0-IV1

ijuma committed:
Remove no longer relevant test

ijuma committed:
Add. back modified test

ijuma committed:
Remove unused method parameter

ijuma commented:
I removed the relevant IBP constants too and fixed the known test failures. I'll check the full test results tomorrow and will do a file by file pass then too.

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1085,17 +1084,14 @@ object GroupMetadataManager {
-      // until a tagged field is introduced or the version is bumped.
-      else 3.toShort
+      if (offsetAndMetadata.expireTimestampMs.isPresent) 1.toShort
+      else maxVersion
```
Is there a need to pass it as an argument to the method or could we just hardcode it?

ijuma replied:
This is needed for some tests that verify the ability to read messages encoded with the old format. I'm not sure if there's a way to make sure there are no messages written with the old format by the time the broker is upgraded to 4.0. Given the usual retention rules, it should be fine, but not sure if we can be sure. Thoughts?

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1112,21 +1108,13 @@ object GroupMetadataManager {
-      // until a tagged field is introduced or the version is bumped.
-      else 3.toShort
-
+                         version: Short = 3): Array[Byte] = {
```
ditto.

ijuma replied:
See the reply to the other comment.

ijuma committed:
Fix single test that failed

ijuma commented:
There was a single failing test - pushed a simple fix.

ijuma committed:
Minor tweak to `GroupMetadataManager.offsetCommitValue`

ijuma commented on core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala:
```diff
@@ -2593,45 +2574,12 @@ class GroupMetadataManagerTest {
-  }
-
-  @Test
-  def testSerdeOffsetCommitValueWithNoneExpireTimestamp(): Unit = {
```
This basically duplicates `testSerdeOffsetCommitValue` and hence I removed it.

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kafka-18465-remove-metadata-version-methods-for-versions-older-than-3.0

* apache-github/trunk:
  KAFKA-18340: Change Dockerfile to use log4j2 yaml instead log4j properties (#18378)
  MINOR: fix flaky RemoteLogManagerTest#testStopPartitionsWithDeletion (#18474)
  KAFKA-18311: Enforcing copartitioned topics (4/N) (#18397)
  KAFKA-18308; Update CoordinatorSerde (#18455)
  KAFKA-18440: Convert AuthorizationException to fatal error in AdminClient (#18435)
  KAFKA-17671: Create better documentation for transactions (#17454)
  KAFKA-18304; Introduce json converter generator (#18458)
  MINOR: Clean up classic group tests (#18473)
  KAFKA-18399 Remove ZooKeeper from KafkaApis (2/N): CONTROLLED_SHUTDOWN and ENVELOPE (#18422)
  MINOR: improve StreamThread periodic processing log (#18430)

ijuma commented on core/src/test/scala/unit/kafka/server/AbstractFetcherThreadTest.scala:
```diff
@@ -390,43 +390,6 @@ class AbstractFetcherThreadTest {
   }
 
-  @Test
-  def testTruncateToHighWatermarkIfLeaderEpochRequestNotSupported(): Unit = {
```
No longer relevant since leader epoch request is now always supported.

ijuma commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -486,74 +473,6 @@ class AlterPartitionManagerTest {
 
-  @ParameterizedTest
-  @MethodSource(Array("provideMetadataVersions"))
-  def testPartialTopicIds(metadataVersion: MetadataVersion): Unit = {
```
We now always have topic ids.

ijuma commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1640,15 +1639,6 @@ class KafkaConfigTest {
   }
 
-  @Test
-  def testInvalidInterBrokerProtocolVersionKRaft(): Unit = {
```
The baseline version now supports kraft.

ijuma commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -130,96 +128,12 @@ class ReplicaFetcherThreadTest {
 
-  @Disabled("KAFKA-18370")
-  @Test
-  def testFetchLeaderEpochRequestIfLastEpochDefinedForSomePartitions(): Unit = {
```
Lots of disabled tests that rely on the "no truncate on fetch", which is no longer applicable.

junrao replied:
It's probably better to remove those tests related OffsetForLeaderEpochRequest until we remove the code in `AbstractFetcherThread.truncateToEpochEndOffsets()`. This can be done in a separate jira.

We could remove all tests that depend on old IBPs.

ijuma replied:
Sorry @junrao - I didn't fully understand. Are you saying we should _not_ remove these tests just yet or that it's ok to remove these now and remove the rest in a separate PR?

junrao replied:
KAFKA-18370 says that we could remove these tests because 4.0 doesn't support ZK anymore. This seems incorrect since this part of the code is independent of the ZK change. However, it does seem that the OffsetForLeaderEpochRequest is not really exercised because of the truncating on fetch support. I am just not 100% sure. So, it would useful to try to remove that part of the code first and then remove the corresponding tests.

cc @chia7712

ijuma replied:
I see. The reason I removed these tests is:

1. They relied on IBP 2.6-IV0 via the `kafkaConfigNoTruncateOnFetch` method.
2. They were disabled.
3. I was under the impression that the truncation approach from IBP 2.7 onwards does not rely on offset for leader epochs requests: #9382

So, I wasn't sure how to make the tests work. 

Your point is that the legacy code is still there and hence removing the tests is premature. Is that right?

ijuma replied:
Since these tests were already disabled, is it OK if I keep them deleted and follow up with a separate change that removes the supposed replica fetcher dead code?

junrao replied:
Sounds good. I filed https://issues.apache.org/jira/browse/KAFKA-18477 to track the removal work in replica fetcher.

ijuma commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -1551,40 +1514,14 @@ public void testNonIncreasingOffsetRecordBatchHasMetricsLogged() {
     }
 
-    @Test
-    public void testZStdCompressedWithUnavailableIBPVersion() {
```
zstd is now always available.

ijuma commented:
This is ready for review now.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/AlterPartitionRequest.java:
```diff
@@ -68,14 +68,12 @@ public static class Builder extends AbstractRequest.Builder<AlterPartitionReques
          *             topics have a topic id and a topic name set.
-         * @param canUseTopicIds True if version 2 and above can be used.
          */
-        public Builder(AlterPartitionRequestData data, boolean canUseTopicIds) {
```
v2 request is added in 3.3.0. To be consistent, we need to move the MV baseline to 3.3.

ijuma replied:
The claim is that we're using protocol negotiation to decide the version to use. The following is stated in `AlterPartitionManager.buildRequest`.

* While building the request, we don't know which version of the AlterPartition API is
* supported by the controller. The final decision is taken when the AlterPartitionRequest
* is built in the network client based on the advertised api versions of the controller.

ijuma replied:
And the code does seem to use `clientChannelBuilder`.

junrao replied:
Thanks for the explanation. We can keep the change as it is then.

junrao commented on core/src/main/scala/kafka/server/ApiVersionManager.scala:
```diff
@@ -150,14 +150,12 @@ class DefaultApiVersionManager(
     }
     val apiVersions = if (controllerApiVersions.isDefined) {
       ApiVersionsResponse.controllerApiVersions(
-        finalizedFeatures.metadataVersion().highestSupportedRecordVersion,
```
Should we leave this param here to support record format change in the future?

ijuma replied:
I can revert this (and related code) if you think it's useful since it's reasonably self-contained. The benefit of keeping vs removing is probably borderline given that there are no concrete plans for adding new record versions (including the specifics regarding the compatibility story). Still, this is a bit unrelated to the main point of this PR (MV/IBP), so I can change it back.

junrao replied:
Yes, I agree that it's borderline. It's probably ok to keep this change for now.

ijuma commented:
The tests passed.

junrao commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1112,21 +1109,13 @@ object GroupMetadataManager {
    * @param assignment the assignment for the rebalancing generation
-   * @param metadataVersion the api version
+   * @param version the version to serialize it with, the default is `3`, the highest supported non-flexible version
+   *        until a tagged version is bumped. The default should always be used outside of tests
```
until a tagged version is bumped => until a tagged field is introduced or the version is bumped

junrao commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/classic/ClassicGroup.java:
```diff
@@ -1445,7 +1445,7 @@ public void createClassicGroupRecords(
         );
 
-        records.add(GroupCoordinatorRecordHelpers.newGroupMetadataRecord(this, assignments, metadataVersion));
+        records.add(GroupCoordinatorRecordHelpers.newGroupMetadataRecord(this, assignments));
```
`metadataVersion` is no longer used.

junrao commented on core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:
```diff
@@ -128,7 +127,7 @@ class LogLoaderTest {
         logDirFailureChannel = logDirFailureChannel,
         time = time,
-        keepPartitionMetadataFile = config.usesTopicId,
+        keepPartitionMetadataFile = true,
```
We can probably remove keepPartitionMetadataFile from LogManager. This can be done in a followup PR.

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -48,7 +46,7 @@ import org.junit.jupiter.params.ParameterizedTest
 import org.mockito.ArgumentCaptor
 import org.mockito.ArgumentMatchers.{any, anyBoolean, anyLong}
-import org.mockito.Mockito.{mock, never, times, verify, when}
+import org.mockito.Mockito.{mock, times, verify, when}
```
We could delete `shouldUseLeaderEndOffsetIfInterBrokerVersionBelow20`.

junrao reviewed (commented):
@ijuma : Thanks for the PR. Left a few comments.

ijuma committed:
Address some of the review comments

ijuma committed:
Fix comment.

ijuma commented:
@junrao I have addressed your comments. Let me know if you have any other comments or if this looks good. Thanks!

junrao reviewed (approved):
@ijuma : Thanks for the updated PR. LGTM.

Filed https://issues.apache.org/jira/browse/KAFKA-18479 to track the followup item.
